src/freertos_drivers/esp_idf/EspIdfWiFi.cxx

```diff
@@ -0,0 +1,1355 @@
+            break;
+        case WIFI_EVENT_AP_STOP:
+            ipAcquiredAp_ = false;
+            ip_acquired(IFACE_AP, false);
+            break;
+        case WIFI_EVENT_AP_STACONNECTED:
+        {
+            wifi_event_ap_staconnected_t *event =
+                (wifi_event_ap_staconnected_t *)data;
+            LOG(INFO, "wifi: station " MACSTR " join, AID=%d",
+                MAC2STR(event->mac), event->aid);
+            ++apClientCount_;
+            break;
+        }
+        case WIFI_EVENT_AP_STADISCONNECTED:
+        {
+            wifi_event_ap_stadisconnected_t *event =
+                (wifi_event_ap_stadisconnected_t *)data;
+            LOG(INFO, "wifi: station " MACSTR " leave, AID=%d",
+                MAC2STR(event->mac), event->aid);
+            if (--apClientCount_ == 0)
+            {
+                ipLeased_ = false;
+            }
+            break;
+        }
+    }
+}
+
+//
+// EspIdfWiFi::ip_event_handler()
+//
+void EspIdfWiFi::ip_event_handler(esp_event_base_t base, int32_t id, void *data)
+{
+    if (base != IP_EVENT)
+    {
+        LOG(WARNING, "wifi: ip_event_handler(), not a IP_EVENT.");
+        return;
+    }
+
+    switch (id)
+    {
+        default:
+            LOG(INFO, "wifi: ip_event_handler() unknown id: %i", (int)id);
+            break;
+        case IP_EVENT_STA_GOT_IP:
+        {
+            {
+                // We have to take a mutex because for some reason the
+                // IP_EVENT_STA_LOST_IP event does not get delivered when there
+                // is an unexpected WiFi disconnect. I guess it is implied?
+                OSMutexLock locker(this);
+                HASSERT(ipAcquiredSta_ == false);
+                ipAcquiredSta_ = true;
+                mdns_restore_sta();
+            }
+            ip_event_got_ip_t *d = static_cast<ip_event_got_ip_t *>(data);
+            char ip_addr[16];
+            inet_ntoa_r(d->ip_info.ip.addr, ip_addr, 16);
+            LOG(INFO, "wifi: STA got IP: %s", ip_addr);
+            ip_acquired(IFACE_STA, true);
+            break;
+        }
+        case IP_EVENT_STA_LOST_IP:
+        {
+            {
+                // We have to take a mutex because for some reason the
+                // IP_EVENT_STA_LOST_IP event does not get delivered when there
+                // is an unexpected WiFi disconnect. I guess it is implied?
+                OSMutexLock locker(this);
+                if (ipAcquiredSta_)
+                {
+                    ipAcquiredSta_ = false;
+                    mdns_disable_sta();
+                }
+            }
+            LOG(INFO, "wifi: STA lost IP.");
+            ip_acquired(IFACE_STA, false);
+            break;
+        }
+        case IP_EVENT_AP_STAIPASSIGNED:
+            ipLeased_ = true;
+            break;
+    }
+}
+
+//
+// EspIdfWiFi::init_config()
+//
+void EspIdfWiFi::init_config()
+{
+    esp_err_t result = nvs_open(NVS_NAMESPACE_NAME, NVS_READWRITE, &cfg_);
```

Comment:
we may need to have a call to `nvs_flash_init` which can return `ESP_ERR_NVS_NO_FREE_PAGES` which we need to handle ourselves as erase the NVS partition to force reinit. This is critical if the user has not called it previously from their code.


We should also have a call to `nvs_close` for every `nvs_open` call made, this should be the final call after any adjustments are done. We should not rely on nvs being open at all times and instead only open/close when needed.

Reply:
looks much cleaner.